bookserver/routers/rslogging.py

```diff
@@ -179,3 +195,135 @@ async def runlog(request: Request, response: Response, data: LogRunIncoming):
 async def same_class(user1: AuthUserValidator, user2: str) -> bool:
     u2 = await fetch_user(user2)
     return user1.course_id == u2.course_id
+
+
+# completion tables
+# =================
+#
+# This section contains implementations of endpoints for tracking progress
+
+
+# updatelastpage
+# --------------
+# see :ref:`processPageState`
+@router.post("/updatelastpage")
+async def updatelastpage(request: Request, request_data: LastPageDataIncoming):
+    if request_data.last_page_url is None:
+        return  # todo:  log request_data, request.args and request.env.path_info
+    if request.state.user:
+        lpd = request_data.dict()
+        rslogger.debug(f"{lpd=}")
+
+        lpd["last_page_chapter"] = request_data.last_page_url.split("/")[-2]
+        lpd["last_page_subchapter"] = ".".join(
+            request_data.last_page_url.split("/")[-1].split(".")[:-1]
+        )
+        lpd["last_page_accessed_on"] = datetime.utcnow()
+        lpd["user_id"] = request.state.user.id
+
+        lpdo: LastPageData = LastPageData(**lpd)
+        await update_user_state(lpdo)
+        await update_sub_chapter_progress(lpdo)
+        # The components don't ever look at a result from this
```

Comment:
I agree; we should also check the returned value from endpoints in the JS for error / success and inform the user.